beacon-chain/sync/data_columns_sampling.go

```diff
@@ -273,132 +294,188 @@ func (s *Service) incrementalDAS(
 			return false, nil, errors.New("retrieved more columns than requested")
 		}
 
-		// Some columns are missing, we need to extend the sample size.
-		missingColumnsCount += columnsToSampleCount - retrievedSampleCount
-
-		firstColumnToSample = extendedSampleCount
+		// missing columns, extend the samples.
+		allowedFailures += columnsToSampleCount - retrievedSampleCount
 		oldExtendedSampleCount := extendedSampleCount
-		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, missingColumnsCount)
+		firstColumnToSample = extendedSampleCount
+		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, allowedFailures)
 
 		log.WithFields(logrus.Fields{
 			"root":                fmt.Sprintf("%#x", root),
 			"round":               round,
-			"missingColumnsCount": missingColumnsCount,
-			"currentSampleCount":  oldExtendedSampleCount,
-			"nextSampleCount":     extendedSampleCount,
+			"missingColumnsCount": allowedFailures,
+			"currentSampleIndex":  oldExtendedSampleCount,
+			"nextSampleIndex":     extendedSampleCount,
 		}).Debug("Some columns are still missing after sampling this round.")
 	}
 }
 
-// DataColumnSamplingRoutine runs incremental DAS on block when received.
-func (s *Service) DataColumnSamplingRoutine(ctx context.Context) {
-	// Get the custody subnets count.
-	custodySubnetsCount := peerdas.CustodySubnetCount()
-
-	// Create a subscription to the state feed.
-	stateChannel := make(chan *feed.Event, 1)
-	stateSub := s.cfg.stateNotifier.StateFeed().Subscribe(stateChannel)
+func (d *dataColumnSampler1D) sampleDataColumns(
+	ctx context.Context,
+	root [fieldparams.RootLength]byte,
+	columns []uint64,
+) (map[uint64]bool, error) {
+	// distribute samples to peer
+	peerToColumns, err := d.distributeSamplesToPeer(columns)
+	if err != nil {
+		log.WithError(err).Error("Failed to distribute samples to peers")
+		return nil, errors.Wrap(err, "failed distributing samples to peer")
+	}
 
-	// Unsubscribe from the state feed when the function returns.
-	defer stateSub.Unsubscribe()
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	res := make(map[uint64]bool)
```

Comment:
We can optimistically hope that we'll successfully sample all the requested columns.
So the capacity of this map can directly be set to `len(columns)`, in order to avoid additional malloc during iteration on `retrieved`.

Reply:
I feel it is a little bit over optimization 😄, the time taken on these would be negligible here.

Reply:
For normal cases, I would agree if we have determined size of the map and if they're large.

But with number of columns and samples in the double digit range, not defining sizes seems totally fine with almost no performance penalties

Reply:
> not defining sizes seems totally fine with almost no performance penalties

With defined size, there is no performance penalty at all.
I see only advantages to define the size.
What is the advantage **not** to define it?

Reply:
I feel this is over optimization, creating maps with / without size won't impact the performance materially especially it's like ~100 or less entries.